File: examples/demo7.m
if ~libisloaded('mutilities')
    fprintf('loading mutilities library...\n')
    [notfound, warnings] = loadlibrary('mutilities');
end
if ~libisloaded('mgadgetron')
    fprintf('loading mgadgetron library...\n')
    [notfound, warnings] = loadlibrary('mgadgetron');
end

try

    % define raw data source
    file = input('raw data file: ', 's');
    input_data = gadgetron.MR_Acquisitions(file);

    % pre-process acquisitions
    prep_gadgets = [{'NoiseAdjustGadget'} {'AsymmetricEchoGadget'} ...
         {'RemoveROOversamplingGadget'}];
    acq_proc = gadgetron.AcquisitionsProcessor(prep_gadgets);
    fprintf('---\n pre-processing acquisitions...\n')
    preprocessed_data = acq_proc.process(input_data);

    % perform reconstruction
    recon = gadgetron.MR_BasicGRAPPAReconstruction();
    recon.set_input(preprocessed_data);
    fprintf('---\n reconstructing...\n');
    recon.process();
    % for undersampled acquisition data GRAPPA computes G-factors
    % in addition to reconstructed images
    output = recon.get_output();

    % get real-valued reconstructed images and G-factors
    images = gadgetron.MR_extract_real_images(output.select(2));
    gfacts = gadgetron.MR_extract_real_images(output.select(2,1));

    n = images.number();
    fprintf('Enter slice number to view its data\n')
    fprintf('(a value outside the range [1 : %d] will stop this loop).\n', n)
    while (true)
        i = input('slice: ');
        if i < 1 || i > n
            break
        end
        idata = images.image_as_array(i);
        gdata = gfacts.image_as_array(i);
        idata = idata/max(max(max(idata)));
        gdata = gdata/max(max(max(gdata)));
        figure(i)
        imshow(idata(:,:,1));
        title(['image ' num2str(i)])
        figure(i + n)
        imshow(gdata(:,:,1));
        title(['G-factor ' num2str(i)])
    end
    
catch err
    % display error information
    fprintf('%s\n', err.message)
    fprintf('error id is %s\n', err.identifier)
end

File: examples/demo7.py
'''
Upper-level demo, GRAPPA reconstruction of undersampled data.
'''

import argparse
import os
import pylab
import sys

BUILD_PATH = os.environ.get('BUILD_PATH') + '/xGadgetron'
SRC_PATH = os.environ.get('SRC_PATH') + '/xGadgetron/pGadgetron'

sys.path.append(BUILD_PATH)
sys.path.append(SRC_PATH)

from pGadgetron import *

parser = argparse.ArgumentParser(description = \
'''
Upper-level demo, GRAPPA reconstruction of undersampled data.
''')
parser.add_argument\
('filename', nargs='?', default = 'testdata.h5', \
 help = 'raw data file name (default: testdata.h5)')
args = parser.parse_args()                                 

def main():

    # acquisitions will be read from an HDF file args.filename
    input_data = MR_Acquisitions(args.filename)

    # pre-process acquisitions
    prep_gadgets = ['NoiseAdjustGadget', 'AsymmetricEchoGadget', \
         'RemoveROOversamplingGadget']
    acq_proc = AcquisitionsProcessor(prep_gadgets)
    print('---\n pre-processing acquisitions...')
    preprocessed_data = acq_proc.process(input_data)

    # perform reconstruction
    recon = MR_BasicGRAPPAReconstruction()
    recon.set_input(preprocessed_data)
    print('---\n reconstructing...')
    recon.process()
    output = recon.get_output()
    # for undersampled acquisition data GRAPPA computes Gfactor images
    # in addition to reconstructed ones
    complex_images = output.select(2)
    complex_gfactors = output.select(2, 1)

    # get real-valued reconstructed images and gfactors
    print('---\n processing images...')
    images = MR_extract_real_images(complex_images)
    gfactors = MR_extract_real_images(complex_gfactors)

    nz = images.number()
    print('%d images reconstructed.' % nz)

    # plot images and gfactors
    print('Enter the number of the slice to view it')
    print('(a value outside the range [1 : %d] will stop this loop)'% nz)
    while True:
        s = str(input('slice: '))
        if len(s) < 1:
            break
        z = int(s)
        if z < 1 or z > nz:
            break
        data = images.image_as_array(z - 1)
        gdata = gfactors.image_as_array(z - 1)
        pylab.figure(z)
        pylab.title('image')
        pylab.imshow(data[0,0,:,:])
        print('Close Figure %d window to continue...' % z)
        pylab.figure(z + nz)
        pylab.title('G factor')
        pylab.imshow(gdata[0,0,:,:])
        print('Close Figure %d window to continue...' % (z + nz))
        pylab.show()

try:
    main()
    print('done')

except error as err:
    # display error information
    print ('Gadgetron exception occured:\n', err.value)
